# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/base/callback_internal.cc
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 3
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/base/callback_internal.cc
--- a/security/sandbox/chromium/base/callback_internal.cc	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/base/callback_internal.cc	Wed Jun 12 11:10:48 2019 +0100
@@ -11,7 +11,16 @@
 
 namespace {
 
-bool ReturnFalse(const BindStateBase*) {
+bool QueryCancellationTraitsForNonCancellables(
+    const BindStateBase*,
+    BindStateBase::CancellationQueryMode mode) {
+  switch (mode) {
+    case BindStateBase::IS_CANCELLED:
+      return false;
+    case BindStateBase::MAYBE_VALID:
+      return true;
+  }
+  NOTREACHED();
   return false;
 }
 
@@ -23,18 +32,20 @@
 
 BindStateBase::BindStateBase(InvokeFuncStorage polymorphic_invoke,
                              void (*destructor)(const BindStateBase*))
-    : BindStateBase(polymorphic_invoke, destructor, &ReturnFalse) {
-}
+    : BindStateBase(polymorphic_invoke,
+                    destructor,
+                    &QueryCancellationTraitsForNonCancellables) {}
 
-BindStateBase::BindStateBase(InvokeFuncStorage polymorphic_invoke,
-                             void (*destructor)(const BindStateBase*),
-                             bool (*is_cancelled)(const BindStateBase*))
+BindStateBase::BindStateBase(
+    InvokeFuncStorage polymorphic_invoke,
+    void (*destructor)(const BindStateBase*),
+    bool (*query_cancellation_traits)(const BindStateBase*,
+                                      CancellationQueryMode))
     : polymorphic_invoke_(polymorphic_invoke),
       destructor_(destructor),
-      is_cancelled_(is_cancelled) {}
+      query_cancellation_traits_(query_cancellation_traits) {}
 
-CallbackBase::CallbackBase(CallbackBase&& c) = default;
-CallbackBase& CallbackBase::operator=(CallbackBase&& c) = default;
+CallbackBase& CallbackBase::operator=(CallbackBase&& c) noexcept = default;
 CallbackBase::CallbackBase(const CallbackBaseCopyable& c)
     : bind_state_(c.bind_state_) {}
 
@@ -43,10 +54,10 @@
   return *this;
 }
 
-CallbackBase::CallbackBase(CallbackBaseCopyable&& c)
+CallbackBase::CallbackBase(CallbackBaseCopyable&& c) noexcept
     : bind_state_(std::move(c.bind_state_)) {}
 
-CallbackBase& CallbackBase::operator=(CallbackBaseCopyable&& c) {
+CallbackBase& CallbackBase::operator=(CallbackBaseCopyable&& c) noexcept {
   bind_state_ = std::move(c.bind_state_);
   return *this;
 }
@@ -62,24 +73,21 @@
   return bind_state_->IsCancelled();
 }
 
+bool CallbackBase::MaybeValid() const {
+  DCHECK(bind_state_);
+  return bind_state_->MaybeValid();
+}
+
 bool CallbackBase::EqualsInternal(const CallbackBase& other) const {
   return bind_state_ == other.bind_state_;
 }
 
-CallbackBase::CallbackBase(BindStateBase* bind_state)
-    : bind_state_(bind_state ? AdoptRef(bind_state) : nullptr) {
-  DCHECK(!bind_state_.get() || bind_state_->HasOneRef());
-}
+CallbackBase::~CallbackBase() = default;
 
-CallbackBase::~CallbackBase() {}
-
-CallbackBaseCopyable::CallbackBaseCopyable(const CallbackBaseCopyable& c)
-    : CallbackBase(nullptr) {
+CallbackBaseCopyable::CallbackBaseCopyable(const CallbackBaseCopyable& c) {
   bind_state_ = c.bind_state_;
 }
 
-CallbackBaseCopyable::CallbackBaseCopyable(CallbackBaseCopyable&& c) = default;
-
 CallbackBaseCopyable& CallbackBaseCopyable::operator=(
     const CallbackBaseCopyable& c) {
   bind_state_ = c.bind_state_;
@@ -87,7 +95,7 @@
 }
 
 CallbackBaseCopyable& CallbackBaseCopyable::operator=(
-    CallbackBaseCopyable&& c) = default;
+    CallbackBaseCopyable&& c) noexcept = default;
 
 }  // namespace internal
 }  // namespace base